src/hera/workflows/resource_base.py

```diff
@@ -0,0 +1,197 @@
+"""Define the base classes used to produce k8s CRDs."""
+from __future__ import annotations
+
+from typing import Any, Callable, ClassVar, Dict, List, Optional, Type, Union
+
+try:
+    from inspect import get_annotations  # type: ignore
+except ImportError:
+    from hera.workflows._inspect import get_annotations  # type: ignore
+from pathlib import Path
+
+try:
+    from typing import Annotated, get_args, get_origin  # type: ignore
+except ImportError:
+    from typing_extensions import Annotated, get_args, get_origin  # type: ignore
+
+try:
+    from typing import Self  # type: ignore
+except ImportError:
+    from typing_extensions import Self  # type: ignore
+
+from hera import _yaml
+from hera.shared import BaseMixin
+from hera.shared._pydantic import PydanticBaseModel, get_fields
+
+
+class ModelMapper:
+    """Maps a model attribute to the attribute path of the produced resource.
+
+    For example `foo: MopdelMapper("metadata.name")` would map the annotated object's
+    `foo` attribute to the "metadata" -> "name" field on the final k8s resource.
+    """
+
+    def __init__(self, model_path: str, hera_builder: Optional[Callable] = None):
+        """Accept the `model_path` to map to."""
+        self.model_path = None
+        self.builder = hera_builder
+
+        if not model_path:
+            # Allows overriding parent attribute annotations to remove the mapping
+            return
+
+        self.model_path = model_path.split(".")
+
+    @classmethod
+    def build_model(
+        cls,
+        hera_class: Type[ResourceBase],
+        hera_obj: ResourceBase,
+        model: PydanticBaseModel,
+        *,
+        traverse_mro: bool = True,
+    ):
+        """Builds an instance of a `ResourceBase` into the destination model type."""
+        assert isinstance(hera_obj, ResourceBase), type(hera_obj)
+
+        for attr, mapper, _ in hera_class._iter_model_mappers(traverse_mro=traverse_mro):
+            # Value comes from builder function if it exists on hera_obj, otherwise directly from the attr
+            value = (
+                getattr(hera_obj, mapper.builder.__name__)() if mapper.builder is not None else getattr(hera_obj, attr)
+            )
+            if value is not None:
+                _set_model_attr(model, mapper.model_path, value)
+
+        return model
+
+
+class ResourceBase(BaseMixin):
+    """Base class for top-level Hera CRDs.
+
+    Subclasses should supply a `mapped_model` attribute, which is the Pydantic model
```

Comment:
This should probably be a private attribute, I'm curious why you changed from a mapper subclass used in the annotation to a `mapped_model` attribute?

Reply:
you mean the
```
class _WorkflowModelMapper(_ModelMapper):
    @classmethod
    def _get_model_class(cls) -> Type[BaseModel]:
        return _ModelWorkflow
```

pattern?

If so, it seemed like a very roundabout way of asserting that all the mappers for a given source class should use the same mapper.

* There was nothing ensuring an author actually used the subclass
* it was a decent amount of boilerplate. less important internally. but this change kind of intentionally makes this a public facing detail now.
* The logic associated with each subclass was the same, it seemed like the only purpose was to specify the desitnation model to map to.

---

By contrast (i think) having it be a class attribute helps in a few ways:

* It's an attribute of the model, so it automatically applies the same destination class to all attributes
* less boilerplate, 1 line configuration for a source class
* validation happens at class creation time. I personally think this is nicer than happening at `_WorkflowModelMapper` instantiation time. you have more information about the class being mapped for error messages and such

Reply:
as for it being a private attribute, it would seem (to me) to be weird for it to be a private attribute. An external definer of a CRD would be required to make use of this field

```python
class MyCRD(ResourceBase):
    mapped_model: ClassVar[Type] = MyCRDModel

    foo: Annotated[str, ModelMapper('foo')]

class MyCRDModel(BaseModel):
    foo: str
```

Reply:
I think my intention about `mapped_model` being private was more the intention that you shouldn't be able to (and shouldn't need to) change it as it's an intrinsic property of your `ResourceBase`, so maybe a property would be more appropriate? Otherwise I can do this when defining a `Workflow`:
```py
with Workflow(mapped_model=IrrelevantModel):
    ...
```

Reply:
It's a `ClassVar`, which will make pydantic ignore it as far as input parameters go. I'd sooner make it `_mapped_model` than make it a property personally, but I still dont personally think it **needs** to be private.

I could also rename it to something like `hera_model` or something so it's less likely to conflict with arbitrary downstream code.

Reply:
It occurred to me that it might be...something idk...a viable alternative to define it as `__init_subclass__` arguments like `class Workflow(ResourceBase, mapped_model=_WorkflowModel):`

It does cut down on the necessary typing a little bit, and it pulls it out of the class body where, as your prior comments implied, it could be confused as a model input param.

However, `__init_subclass__` is basically opting out of type enforcement on that argument, and the value ultimately would need to be stashed on the type anyways. So it ends up being kind of a wash.